File: ver-android/app/src/main/kotlin/org/opengamestudio/budget/BudgetView.kt
package org.opengamestudio

import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.unit.dp

@Composable
fun BudgetView(
    modifier: Modifier = Modifier,
    vm: VM,
) {
    Column(
        modifier = modifier
            .fillMaxSize()
            .padding(top = 8.dp)
            .padding(horizontal = 24.dp)
    ) {
        Row {
            TextField(
                keyboardOptions = KeyboardOptions(imeAction = ImeAction.Next),
                modifier = Modifier.weight(2f),
                onValueChange = { budgetSet(F.inputSpent, it) },
                placeholder = { Text(vm.spentPlaceholder.value) },
                value = vm.spent.value,
            )
            Spacer(Modifier.width(8.dp))
            FilledTonalButton(
                modifier = Modifier.weight(1f),
                onClick = { budgetSet(F.didClickPasteSpent, true) },
            ) {
                Text(vm.spentPasteTitle.value)
            }
        }
        Spacer(Modifier.height(8.dp))
        Row {
            TextField(
                keyboardOptions = KeyboardOptions(imeAction = ImeAction.Done),
                modifier = Modifier.weight(2f),
                onValueChange = { budgetSet(F.inputMorningBalance, it) },
                placeholder = { Text(vm.morningBalancePlaceholder.value) },
                value = vm.morningBalance.value,
            )
            Spacer(Modifier.width(8.dp))
            FilledTonalButton(
                modifier = Modifier.weight(1f),
                onClick = { budgetSet(F.didClickPasteMorningBalance, true) },
            ) {
                Text(vm.morningBalancePasteTitle.value)
            }
        }
        Spacer(Modifier.height(8.dp))
        Column(modifier = Modifier
            .border(
                color = Color.LightGray,
                shape = RoundedCornerShape(8.dp),
                width = 1.dp
            )
            .padding(8.dp)
            .fillMaxWidth()
        ) {
            Text(
                modifier = Modifier.padding(top = 12.dp),
                style = MaterialTheme.typography.bodyMedium,
                text = vm.result.value,
            )
            Spacer(Modifier.height(8.dp))
            Row(
                horizontalArrangement = Arrangement.End,
                modifier = Modifier.fillMaxWidth()
            ) {
                FilledTonalButton(
                    onClick = { budgetSet(F.didClickCopy, true) },
                ) {
                    Text(vm.resultCopyTitle.value)
                }
            }
        }
        Spacer(Modifier.height(8.dp))
        Row(
            //horizontalArrangement = Arrangement.End,
            modifier = Modifier.fillMaxWidth()
        ) {
            FilledTonalButton(
                onClick = { budgetSet(F.didClickCommit, true) },
            ) {
                Text(vm.commitTitle.value)
            }
            Spacer(modifier = Modifier.fillMaxWidth())
            FilledTonalButton(
                onClick = { budgetSet(F.didClickPush, true) },
            ) {
                Text(vm.pushTitle.value)
            }
        }
    }
}

File: ver-ios/src/budget/BudgetView.swift
import SwiftUI

struct BudgetView: View {
    @ObservedObject var vm: VM

    init(_ vm: VM) {
        self.vm = vm
    }

    var body: some View {
        VStack(spacing: 8) {
            HStack(spacing: 8, alignment: .center) {
                TextField(
                    "",
                    text: Binding(
                        get: { vm.spent },
                        set: { budgetSet(F.inputSpent, $0) }
                    ),
                    prompt: Text(vm.spentPlaceholder)
                )
                    .textFieldStyle(.roundedBorder)
                Button(action: { budgetSet(F.didClickPasteSpent, true) }) {
                    Text(vm.spentPasteTitle)
                }
                    .buttonStyle(.borderedProminent)
            }
            HStack(spacing: 8, alignment: .center) {
                TextField(
                    "",
                    text: Binding(
                        get: { vm.morningBalance },
                        set: { budgetSet(F.inputMorningBalance, $0) }
                    ),
                    prompt: Text(vm.morningBalancePlaceholder)
                )
                    .textFieldStyle(.roundedBorder)
                Button(action: { budgetSet(F.didClickPasteMorningBalance, true) }) {
                    Text(vm.morningBalancePasteTitle)
                }
                    .buttonStyle(.borderedProminent)
            }
            VStack(spacing: 8) {
                HStack {
                    Text(vm.result)
                    Spacer()
                }
                HStack {
                    Spacer()
                    Button(action: { budgetSet(F.didClickCopy, true) }) {
                        Text(vm.resultCopyTitle)
                    }
                        .buttonStyle(.borderedProminent)
                }
            }
                .padding()
                .background(
                    RoundedRectangle(cornerRadius: 8)
                        .stroke(Color.gray.opacity(0.3), lineWidth: 1)
                )
            Spacer()
        }
            .padding([.leading, .trailing])
    }
}
